modules/SequencesExt.tla

```diff
@@ -124,6 +124,37 @@ Last(s) ==
 
 -----------------------------------------------------------------------------
 
+IsSubSequence(s, t) ==
+  (**************************************************************************)
+  (* TRUE iff the sequence s is a subsequence of the sequence t, s.t.       *)
+  (**************************************************************************)
+  \E i,j \in 1..Len(t):
+    /\ i <= j \* Optimization to not eval SubSeq with j < i, which is defined as <<>>.
+    /\ SubSeq(t, i, j) = s
+
+ReplaceFirstSubSeq(r, s, t) ==
+  (**************************************************************************)
+  (* The sequence  t  with its subsequence  s  replaced by the sequence  r  *)
+  (**************************************************************************)
+  IF IsSubSequence(s, t) THEN \* Prevent CHOOSE x \in S: P not holding for any element.
+    LET idx == CHOOSE i \in 1..Len(t): \E j \in i..Len(t): SubSeq(t, i, j) = s
+        suffix == r \o SubSeq(t, idx + Len(s), Len(t))
+    IN IF idx > 1
+       THEN SubSeq(t, 1, idx - 1) \o suffix
+       ELSE suffix
+  ELSE t
+
+RECURSIVE ReplaceAllSubSeqs(_,_,_)
+ReplaceAllSubSeqs(r, s, t) ==
+  (**************************************************************************)
+  (* The sequence  t  with all subsequences  s  replaced by the sequence  r *)
+  (**************************************************************************)
+  IF IsSubSequence(s, t)
+  THEN ReplaceAllSubSeqs(r, s, ReplaceFirstSubSeq(r, s, t))
```

Comment:
I expect ReplaceAllSubSeqs(r, s, s) to loop, just like ReplaceAllSubSeqs(r, <<>>, t), right? Perhaps add a warning to the comment?